tracer/src/Datadog.Trace/Agent/MessagePack/SpanMessagePackFormatter.cs

```diff
@@ -50,18 +54,70 @@ private SpanMessagePackFormatter()
             _processIdValueBytes = processId > 0 ? MessagePackSerializer.Serialize(processId) : null;
         }
 
+        // IMessagePackFormatter<ArraySegment<Span>> is deprecated, use IMessagePackFormatter<TraceMessageChunk>
+        public int Serialize(ref byte[] bytes, int offset, ArraySegment<Span> value, IFormatterResolver formatterResolver)
+        {
+            var traceContext = value.Array![value.Offset].Context.TraceContext;
+            var traceChunk = new TraceChunkModel(value, traceContext);
+            return Serialize(ref bytes, offset, traceChunk, formatterResolver);
+        }
+
+        public int Serialize(ref byte[] bytes, int offset, TraceChunkModel value, IFormatterResolver formatterResolver)
+        {
+            int originalOffset = offset;
+            var spans = value.Spans;
+            var spanIds = new HashSet<ulong>();
+
+            // Using a for loop to avoid the boxing allocation on ArraySegment.GetEnumerator
+            for (var i = 0; i < spans.Count; i++)
+            {
+                // get all span ids for quick lookup below
+                var span = spans.Array![i + spans.Offset];
+                spanIds.Add(span.SpanId);
+            }
+
+            // start writing span[]
+            offset += MessagePackBinary.WriteArrayHeader(ref bytes, offset, spans.Count);
+
+            // Using a for loop to avoid the boxing allocation on ArraySegment.GetEnumerator
+            for (var i = 0; i < spans.Count; i++)
+            {
+                var span = spans.Array![i + spans.Offset];
+                var parentSpanId = span.Context.ParentId ?? 0;
+                bool isRoot = parentSpanId == 0 || span == span.Context.TraceContext.RootSpan;
+                bool isOrphan = parentSpanId > 0 && !spanIds.Contains(parentSpanId);
+                bool isFirst = i == 0;
+
+                var spanModel = new SpanModel(span, value, isRoot, isOrphan, isFirst);
+                offset += Serialize(ref bytes, offset, spanModel, formatterResolver);
+            }
+
+            return offset - originalOffset;
+        }
+
+        // IMessagePackFormatter<Span> is deprecated, use IMessagePackFormatter<SpanModel>
         public int Serialize(ref byte[] bytes, int offset, Span value, IFormatterResolver formatterResolver)
         {
+            // without more context, we can only determine if this span is a local root
+            bool isRoot = value.Context.ParentId is null or 0 || value == value.Context.TraceContext.RootSpan;
+            var spanModel = new SpanModel(value, traceChunk: null, isRoot, isChunkOrphan: false, isFirstSpanInChunk: false);
```

Comment:
Is `isFirstSpanInChunk` being used anywhere or do you have immediate plans to use it?

Reply:
Technically we're supposed to be doing that already for `_dd.p.dm` but we're not 😅